auth-service/src/main/kotlin/uk/gov/communities/delta/auth/services/PasswordTokenService.kt

```diff
@@ -9,7 +9,33 @@ import java.sql.Timestamp
 import java.time.Instant
 import kotlin.time.Duration.Companion.hours
 
-class SetPasswordTokenService(private val dbPool: DbPool, private val timeSource: TimeSource) {
+class RegistrationSetPasswordTokenService(private val dbPool: DbPool, timeSource: TimeSource) : PasswordTokenService("set_password_tokens", dbPool, timeSource){
+    suspend fun passwordNeverSetForUserCN(userCN: String): Boolean {
+        return withContext(Dispatchers.IO) {
+            setPasswordTokenExistsForUserCN(userCN)
+        }
+    }
+
+    @Blocking
+    private fun setPasswordTokenExistsForUserCN(userCN: String): Boolean {
+        return dbPool.useConnectionBlocking("Check if set password token exists") {
+            val stmt = it.prepareStatement(
+                "SELECT user_cn, token, created_at FROM set_password_tokens " +
+                        "WHERE user_cn = ?"
+            )
+            stmt.setString(1, userCN)
+            val result = stmt.executeQuery()
+            it.commit()
+
+            // Returns true if there is a matching entry, false if not
+            result.next()
+        }
+    }
+}
+
+class ResetPasswordTokenService(dbPool: DbPool, timeSource: TimeSource) : PasswordTokenService("reset_password_tokens", dbPool, timeSource)
+
+abstract class PasswordTokenService(private val tableName: String, private val dbPool: DbPool, private val timeSource: TimeSource) {
```

Comment:
The way you've done it also works, I was imagining making the table name abstract like this which I think is slightly nicer:

```suggestion
class ResetPasswordTokenService(dbPool: DbPool, timeSource: TimeSource) : PasswordTokenService(dbPool, timeSource) {
    override val tableName = "reset_password_tokens"
}

abstract class PasswordTokenService(private val dbPool: DbPool, private val timeSource: TimeSource) {
    protected abstract val tableName: String
```